### ISSUE TITLE
Create a resized copy of an image

### ISSUE THREAD

**Issue by toptensoftware** [dlemstra/Magick.NET#1577]
### Is your feature request related to a problem? Please describe

I have a project that works with very large tif images.  I need to keep the large image in memory for later processing, but I also need smaller versions for on-screen rendering.  Currently to do this I need to duplicate the large image (doubling the memory usage) and then resize it to the smaller size. 

### Describe the solution you'd like

Would like the ability to directly create a resized copy of a source image (or part thereof)

### Describe alternatives you've considered

As mentioned above, the current solution is to create copy of full image and then downsize it - but this requires more memory than directly resizing to a smaller image.

(Perhaps there's already a way to do this but if so, I couldn't figure out how)

**Comment by dlemstra** [comment]
In theory this could be possible for methods of ImageMagick that return a new image. It's an interesting idea but I will need to think about an API for this. And that will probably happen after I have added a new source code generator to call the native api.

**Comment by toptensoftware** [comment]
Sounds good, I look forward to it.

**Comment by dlemstra** [comment]
Do you have any ideas how I could/should integrate this in the API? I think this behavior would only be required for methods that change the dimensions of an image. Otherwise you could just use `.Clone()` and change the image. I am now thinking about something like this:

```C#
using var image = new MagickImage("input.png");
using var thumbnail = image.NewMethod().Resize(100, 100);
using var thumbnail = image.NewProperty.Resize(100, 100);
```
The `NewMethod()` or `NewProperty` would return an interface that includes a the methods that change the dimensions of the image. But the disadvantage of this would be the discoverability in the api. So I am also thinking about something like this:

```C#
using var image = new MagickImage("input.png");
using var thumbnail = image.ResizeWithSuffix(100, 100);
```

As you can see besides deciding which option to use the naming is also something that I have to think about. Do you have any suggestions?

**Comment by toptensoftware** [comment]
Not sure how feasible this is, and I realize this is adding to the original requirement, but the ideal for me would be something like this:

```cs
IMagickImage<QuantumType> CloneCropAndResize(IMagickGeometry? crop, IMagickGeometry? resize)
```

Where:

* `crop` specifies which portion of the source image to grab.  If null, the entire source image
* `resize` specifies how to resize it into a new image instance.  If null, no resizing.

ie: it would be the equivalent of the following but internally would combine all this into a single resize operation without the overhead of the full clone at the start.

```cs
image.Clone().Crop(crop).Resize(resize);
```

The nice thing about this is it would let you grab any region at any resolution. ie: you could grab tiled sections of a source image at display resolutions for rendering.

The name `CloneCropAndResize` is a bit wordy but sums up what it's doing.  Alternative suggestion: `ExtractImage`, or maybe `ExtractSubImage`?

**Comment by dlemstra** [comment]
Those are not the only two methods this option will become available to. Creating a new method for each possible combination is not feasible.

**Comment by toptensoftware** [comment]
Ah ok, I see what you mean.

Perhaps `NewMethod()` could be `CloneThen()`.   

```
var newImage = image.CloneThen().Resize(100, 100);
```

I guess the downside to this is it might be tempting to chain these together fluent style, but the result of the Crop call wouldn't be disposed.

```
var newImage = image
    .CloneThen().Crop(10, 10, 1000, 1000)
    .CloneThen().Resize(100, 100)
```

Thinking out loud (again no idea if this is feasible) but what if you could put together a chain of operations which you terminated with a function to actually do the work. ie: each intermediate operation just returned an interface (not an image instance) upon which you could perform additional operations. The `Process()` call at the end walks the chain of operations and produces a final image.

eg:

```
var newImage = image
    .CloneThen()
    .Crop(10, 10, 1000, 1000)
    .Resize(100, 100)
    .Whatever()
    .SomethingElse()
    .Process();
```

You're obviously thinking about this from a broader perspective than my narrow view... I'm sure whatever you come up with will be fine.

**Comment by alexeygritsenko** [comment]
Good question. But I don’t think that new methods are needed, because it's already supported quite simply:
```
 using MagickImage image = new MagickImage(inputStream);
 using MagickImage thumbnail = image.Clone() as MagickImage;
 thumbnail.Resize(100, 100);
```

**Comment by dlemstra** [comment]
The `Clone()` method creates of copy of the pixels of the input image and then resizes that to the new dimensions. Here is an example what it does with the memory (the numbers are made up but should give you an idea of what happens):

```mermaid
flowchart TD
    A[Create image] --> B[10 mb used]
    B --> C[Clone]
    C --> D[20 mb used]
    D --> E[Resize to 50%]
    B --> F[Resize without clone]
    E --> G[15 mb used]
    F --> G
```

**Comment by dlemstra** [comment]
I now have a working prototype that uses the following syntax:

```
using var image = new MagickImage(Files.Builtin.Logo);
using var clone = image.CloneAndMutate(static mutator => mutator.Resize(100, 100)); // The static keyword is optional
```

What are your thoughts on this @toptensoftware?

**Comment by toptensoftware** [comment]
Hi @dlemstra,

That's pretty nice... I like it.  I assume there will be other mutator operations besides just resize or is that it for now?

Brad

**Comment by dlemstra** [comment]
Other operations will be added but you cannot chain them because there are no memory benefits when doing that so you will need to do this:

```C#
using var croppedImage = image.CloneAndMutate(static mutator => mutator.Crop(10, 10, 1000, 1000));
using var resizedImage = croppedImage.CloneAndMutate(static mutator => mutator.Resize(100, 100));
```

**Comment by alexeygritsenko** [comment]
@dlemstra hi, does your mutator support a sequence of multiple operations or just one?
Example:
`using var thumbnail = image.CloneAndMutate(static mutator => mutator.Crop(10, 10, 1000, 1000).Resize(100, 100));`

**Comment by dlemstra** [comment]
In my current proof of concept the mutator will throw an exception when you execute multiple operations. I have done this to avoid complexity in the mutator. There is no memory benefit if I would allow chaining multiple operations. The `Crop` method always creates a new image. And when I send that image to the Resize method it will also create a new image. And if I would allow multiple operations I would need to cleanup the intermediate image.

**Comment by dlemstra** [comment]
Hey @toptensoftware, I will be going to publish a new release tomorrow or on Monday but I won't be able to include all the operations in the new `IMagickImageCreateOperations` interface that will be used by `.CloneAndMutate`. Moving all these actions will take some time. Are there any operations besides `Resize` that you would want to use right away? If you read this after the release you will probably be going to need to wait for a while.

**Comment by toptensoftware** [comment]
Hi @dlemstra  

No urgency at all on this for me.  I've got a working solution for the moment, and probably won't get a chance to improve it with these new changes imminently.

Brad

**Comment by dlemstra** [comment]
Closing this because the resize operation has been added in the latest release. More operations will be added in the future.

**Comment by IDXGI** [comment]
I encountered this issue when trying to switch from ImageSharp to Magick.NET. I handle relatively large images and frequently perform clone, crop, and resize operations to extract a specific area and scale it to screen size for display. The current interface design has two problems:
1. It can only copy and resize the entire image at the same time, which means that each resize operation applies to the whole image. When I only need a small or medium area to display, this can lead to significant performance loss.
2. It is not possible to set the FilterType (such as Lanczos/Nearest) used during CloneAndMutate(Resize op) without changing the original image.

In ImageSharp, which I used previously, this could be implemented like this:
cropImage = image.Clone(new Configuration { PreferContiguousImageBuffers = true }, x => x.Resize(width: pixelWidth, height: pixelHeight, sampler: KnownResamplers.Lanczos2, sourceRectangle: imageCropRect, targetRectangle: new Rectangle(0, 0, pixelWidth, pixelHeight), compand: false));

Resize() serves a lot of options. In actual testing, when only a small area needs to be extracted, the performance is better. Of course, for small areas, I could first perform CloneArea() and then Resize(), but it is unclear when to use CloneAndMutate or CloneArea+Resize for better performance.

Magick.NET has Bilevel and supports lot of format, so I hope it become better then I can switch here :)
(And encountered ToptenSoftware here, haha, my project uses your RichTextKit, easy to use, sincerely thanks)

**Comment by dlemstra** [comment]
Thanks for sharing your thoughts @IDXGI . I think it would make sense to add an extra overload for the resize method to specify the filter. Could you create a new feature request for this so we don't hijack the conversation here?

And to answer your question that is related to this issue. The diagram earlier in this conversation hopefully explains what happens when you use `CloneAndMutate` instead of doing a `Clone` and then resize it. With `CloneArea` and `Resize`  you will first copy the specified area and then resize it. There is no way to optimize this because you are first extracting part of the image.
